Do dependency checking via a validation action

jadenPete opened the pull request:
You can read more about validation actions here:
https://bazel.build/extending/rules#validation_actions

This should enable dependency checking to be performed concurrently with `phase_singlejar`.

Commit:
Generate OutputGroupInfo in a new phase

Previously, it was generated in the "javainfo" phase, which runs before
the "depscheck" phase. "javainfo" can't run after "depscheck" because
"compile" depends on "javainfo", and "depscheck" can't run before
"javainfo" because it depends on "compile".

Commit:
Made dependency checking a validation action

Commit:
Don't include "phantom inputs" in singlejar actions

This shouldn't be necessary, now that we're using validation actions.

DavidANeil commented:
Have you tested that `--norun_validations` correctly skips all of the validations? (i.e. all of the outputs are correctly segregated)

DavidANeil commented:
For when we pull this into our main repo you will want to experiment with `--experimental_use_validation_aspect` to see if it gives better scheduling.

jadenPete commented:
> Have you tested that `--norun_validations` correctly skips all of the validations? (i.e. all of the outputs are correctly segregated)

Good idea. I just tried building `//dependencies/indirect:used_deps_off_c` from within `tests` without `--norun_validations`, and a `buildozer` message appeared. I also built it with `--norun_validations`, and the message didn't appear, so it looks like we're good!

jadenPete commented:
> For when we pull this into our main repo you will want to experiment with `--experimental_use_validation_aspect` to see if it gives better scheduling.

Will do.

jjudd commented on rules/private/phases/phase_zinc_depscheck.bzl:
```diff
@@ -61,21 +60,16 @@ def phase_zinc_depscheck(ctx, g):
-    if deps_configuration.used == "error":
-        outputs.append(deps_checks["used"])
+        if getattr(deps_configuration, name) == "error":
+            outputs.append(deps_check)
```
If you removed `deps_check` from line 24, then where is `deps_check` coming from?

jadenPete replied:
It should be coming from line 28.

jjudd replied:
Totally missed that. My bad.

Commit:
Test that ScalaCheckDeps outputs aren't inputs to other actions

jjudd committed:
Fix race condition. Also make sure to flush output stream before writing the worker response

There was a race condition where a task could be cancelled with the
worker output streams still set to null, which would cause a null
pointer exception in the worker.

We also weren't always flushing the print stream before responding to
Bazel about the work request. This fixes that.
